### PR TITLE
Change some establish_connection logic

### PR DESCRIPTION
**Pull request by arthurnn**
### Summary

Move `establish_connection` to connection handler.
This give us room to deprecate `establish_connection` on the AR::Base to use the one from the connection handler instead.
### Details

As you know, now the connection handler is not coupled with the model anymore, so establish connection should not be a method in the model.

With this API, you would do something like:

```
ActiveRecord::Base.connection_handler.establish_connection({database: 'foo'}, name: "readonly")
class User < ActiveRecord::Base
  self.connection_specification_name = "readonly"
end

ActiveRecord::Base.connection_handler.establish_connection(:some_key_from_databaseyml, name: "readonly")

ActiveRecord::Base.connection_handler.establish_connection("URL://", name: "readonly")
```

This is backwards compatible as far as I can tell, so it is safe to be backported to 5.0.0
### Next steps (all still up to discussion)
- Make AR internals use `ActiveRecord::Base.connection_handler.establish_connection` API
- Deprecate `AR::Base.establish_connection`
- Make ConnectionSpecification only work with a hash or a string.
  - Hash: when passing a full config
  - String: using a database url
  - Symbol: we wont need this anymore, this is used to point to a config from the databases.yml file (see next item)
- Change database.yml file to the following:

```
production:
  primary: 
    database: foo
  secondary:
    database: foo_readonly
  shard:
    database: shard1
```

Like that, the railties AR initializer, can iterate over the ENV key, and create pools for each of the configs using the right name.

cc @jeremy @tenderlove @rafaelfranca @matthewd 
also @jrafanie as we discussed that today.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 1]
```diff
@@ -180,6 +180,10 @@ def spec(config, name = nil)
 
           adapter_method = "#{spec[:adapter]}_connection"
 
+          unless ActiveRecord::Base.respond_to?(adapter_method)
+            raise AdapterNotFound, "database configuration specifies nonexistent #{spec.config[:adapter]} adapter"
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Think this error message is pretty confusing. What should users do if they hit this error?

**Comment by kaspth** [comment]
This is not the nicest public API:

```ruby
ActiveRecord::Base.connection_handler.establish_connection({database: 'foo'}, name: "readonly")
```

How about making `name` a reserved keyword, then the rest is just the spec:

```ruby
ActiveRecord::Base.connection_handler.establish_connection(name: 'readonly', database: 'foo')
```

**Comment by arthurnn** [comment]
@kaspth thats a wonderful point. I kinda had that in the back of my mind, but was not sure if that would be backwards compatible. I think it will tho, right?

File: activerecord/lib/active_record/connection_handling.rb [hunk 2]
```diff
@@ -44,21 +44,14 @@ module ConnectionHandling
...
       raise "Anonymous class is not allowed." unless name
 
-      spec     ||= DEFAULT_ENV.call.to_sym
-      resolver =   ConnectionAdapters::ConnectionSpecification::Resolver.new configurations
-      # TODO: uses name on establish_connection, for backwards compatibility
-      spec     =   resolver.spec(spec, self == Base ? "primary" : name)
-      self.connection_specification_name = spec.name
+      config ||= DEFAULT_ENV.call.to_sym
+      spec_name = self == Base ? "primary" : name
+      self.connection_specification_name = spec_name
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_handling.rb]: Rereading this, it comes off as needlessly verbose. `specification` is pretty abstract and it feels like we're pushing implementation to the users' eyeballs. Users should only have to think in terms of `connections` — I think we should shorten this to `self.connection_name = 'primary'`.

Comment (kaspth) [inline activerecord/lib/active_record/connection_handling.rb]: Then we could also remove that pesky local variable `spec_name` 🤓

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 3]
```diff
@@ -839,7 +839,15 @@ def connection_pool_list
       end
       alias :connection_pools :connection_pool_list
 
-      def establish_connection(spec)
+      def establish_connection(spec_or_config, name: "primary")
+        if spec_or_config.is_a?(ConnectionSpecification)
+          spec = spec_or_config
+        else
+          resolver = ConnectionAdapters::ConnectionSpecification::Resolver.new(ActiveRecord::Base.configurations)
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Could shorten this a bit by leaving `ConnectionAdapters` and `ActiveRecord` implicit and rely on relative constant lookups 😁

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 4]
```diff
@@ -839,7 +839,15 @@ def connection_pool_list
...
+          spec = resolver.spec(spec_or_config, name)
+        end
+
+        remove_connection(spec.name)
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Why do we need to remove the connection?

Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Ah, that was done in the old `establish_connection`.

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 5]
```diff
@@ -839,7 +839,15 @@ def connection_pool_list
       end
       alias :connection_pools :connection_pool_list
 
-      def establish_connection(spec)
+      def establish_connection(spec_or_config, name: "primary")
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: Think we should leave `name:` a required argument (and make `spec_or_config` optional, if I'm understanding the intent correctly).

Also, I think `spec` and `config` are almost synonyms, which seems to suggest there's something we need to do here. Otherwise that ambiguity bubbles up to users. I'm not sure what that will look like, just saying this makes me a bit uneasy.

**Comment by kaspth** [comment]
@arthurnn as long as the old `establish_connection` passes a name in to the new connection handler version we should be golden, I think.

**Comment by arthurnn** [comment]
@kaspth thanks a lot for reviewing this. i havent much time to work on it. I will this week.

**Comment by kaspth** [comment]
@arthurnn No problem! ❤️

I have another question too. What about YAML extensions in `database.yml`? I think your example above might be a bit too simplistic:

```yaml
production:
  primary: 
    database: foo
  secondary:
    database: foo_readonly
  shard:
    database: shard1
```

For instance taking inspiration from the default `database.yml` we generate it would look like:

```yaml
default: &default
  adapter: sqlite3
  pool: 5
  timeout: 5000

production:
  primary:
    <<: *default
    database: foo
  secondary:
    <<: *default
    database: foo_readonly
  shard:
    <<: *default
    database: shard1
```

Suddenly it doesn't look as clean and I'm wondering whether there's more we should do in steer people in a certain direction with these extra connections.

**Comment by kaspth** [comment]
Looks like @dhh just solved my problem with the syntax above in https://github.com/rails/rails/commit/e530534265d2c32b5c5f772e81cb9002dcf5e9cf#commitcomment-17568231 ✌️

**Comment by bf4** [comment]
This PR looks like it might make it easier for me to use a connection object for our special `DatamartConnection` that just calls stored procures without needing to subclass `ActiveRecord::Base`, no?

**Comment by arthurnn** [comment]
@kaspth thanks for reviewing this. So, I did some changes here.
Mainly, I was able to remove that extra `name` argument I was passing around. I replaced that with the spec hash, we pass around. 
The idea here is to have two different hashes here:
- First the `config` hash, which is just the normal database configuration hash. That can include a `name` which will be used by the Resolver.
- Second the `spec` hash, which will include the `name`, as that will be transformed to an `ConnectionSpecification` object.

So, I changed the `Resolver`, to resolve the config and merge the spec_name.

Let me know what you think ❤️

File: activerecord/lib/active_record/connection_handling.rb [hunk 6]
```diff
@@ -44,21 +44,18 @@ module ConnectionHandling
...
 
-      unless respond_to?(spec.adapter_method)
-        raise AdapterNotFound, "database configuration specifies nonexistent #{spec.config[:adapter]} adapter"
-      end
+      resolver = ConnectionAdapters::ConnectionSpecification::Resolver.new(Base.configurations)
+      spec = resolver.resolve(config).symbolize_keys
+      spec[:name] = spec_name
```
Comment (arthurnn) [inline activerecord/lib/active_record/connection_handling.rb]: We need to `resolve` here, because this method can receive a String, Symbol, or Hash(config).
This is just for backwards compatibility, but eventually the idea is to deprecate this.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 7]
```diff
@@ -13,6 +13,10 @@ def initialize_dup(original)
         @config = original.config.dup
       end
 
+      def to_hash
+        @config.merge(name: @name)
+      end
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Dig this 👍

**Comment by kaspth** [comment]
Can't spot if this already happens, but should we interpret the env key as the spec name?

E.g. this spec will be named production:

```yaml
production:
  database: mah_db
```

Regardless this looks good to me 👍

**Comment by arthurnn** [comment]
@kaspth yeah. this is not done here yet. I will work on that in a next PR, specially because I might change the database.yml file as discussed above.

**Comment by kaspth** [comment]
Sounds good 👍 — just wanted to hear if it was on the table 😁

**Comment by bquorning** [comment]
Do you know if this change will be released with v5.0.0 or v5.1.0?

**Comment by kaspth** [comment]
@bquorning 5.1 definitely, but Rails 5.0 is already in the release candidate phase, so far too late for that.

Because this isn't a bug fix and due to our maintenance policy we won't put this in 5.0.x either. Thanks for the interest! :heart:

**Comment by arthurnn** [comment]
@bquorning i still have some work to do to get the connection management to a better state, that will all be part of 5.1, i will try to start blog about this changes, so people will know about them early on.

**Comment by bquorning** [comment]
Sounds good @arthurnn , @kaspth. I'm just asking because I saw the first part of the refactoring made it into rc1.

**Comment by kaspth** [comment]
No problem 👍  Yeah, that was more by coincidence.

**Comment by schneems** [comment]
I had a thought on the deprecation of `establish_connection`. I think we still need a global way to connect to databases for the case of forking web servers. Currently most everyone recommends putting

```
ActiveRecord::Base.establish_connection
```

In their `after_fork` (or similar) block. If they have to add each individual model to that block, that's going to be painful.

**Comment by arthurnn** [comment]
@schneems me and @matthewd talked the other day, and I think I have a solution to not need the `after_fork` dance anymore. I should implement that as soon as I have some time.(sooner rather than later)

Either way, we could provide some methods like 'connect_all' or something on the connection handler, so you would call them in there instead. Like the disconnect methods we already have in there.
